### PR TITLE
Add vendor/vendorBill record support

### PR DESCRIPTION
**Pull request by SanduDS**
Fixes https://github.com/ballerina-platform/ballerina-extended-library/issues/25
## Goals
Providing more record support through the connector

## Approach
1. Implemented operations related to Vendor records
2. Implemented operations related to Vendor Bill records
3. Implemented  requests with custom xml bodies
4. Implemented search operation to vendor records

## Automation tests
 - Unit tests 
   > Done
 - Integration tests
   > Done

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: netsuite/vendor.bal [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        taxIdNum: extractStringFromXML(response/**/<listRel:taxIdNum>/*),
+        taxItem: extractRecordRefFromXML(response/**/<listRel:taxItem>),
+        password: extractStringFromXML(response/**/<listRel:password>/*),
+        password2: extractStringFromXML(response/**/<listRel:password2>/*),
+        defaultTaxReg: extractRecordRefFromXML(response/**/<listRel:defaultTaxReg>),
+        predictedDays: extractIntegerFromXML(response/**/<listRel:predictedDays>),
+        predConfidence: extractDecimalFromXML(response/**/<listRel:predConfidence>/*)
+    };
+
+    var customFields = extractCustomFiledListFromXML(response/**/<listRel:customFieldList>/*);
```
Comment (LakshanSS) [inline netsuite/vendor.bal:157]: Cant we get rid of `var` here?

File: netsuite/vendor.bal [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    if (customFields is CustomFieldList) {
+        vendor.customFieldList = customFields;
+    }
+
+    boolean|error value = extractBooleanValueFromXMLOrText(response/**/<listRel:isInactive>/*);
```
Comment (LakshanSS) [inline netsuite/vendor.bal:162]: Cant we do like
`boolean value = check extractBooleanValueFromXMLOrText(response/**/<listRel:isInactive>/*);`

So we can get rid of all the `if` blocks below this

File: netsuite/client.bal [hunk 3]
```diff
@@ -40,16 +40,40 @@ public isolated client class Client {
...
+    # + vendorBill - Details of NetSuite record instance creation
+    # + return - RecordAddResponse type record or else the relevant error
+    @display{label: "Add New Vendor Bill"}
+    isolated remote function addNewVendorBill(@display{label: "VendorBill"} NewVendorBill vendorBill) returns 
+                                              @display{label: "Response"} RecordAddResponse|error {
+        xml payload =  check getPayloadAddOp(vendorBill, VENDOR_BILL, self.config);
+        http:Response response = check sendRequest(self.basicClient, ADD_SOAP_ACTION, payload);
+        return getResponseCreationOp(response);
+    }
+
```
Comment (LakshanSS) [inline netsuite/client.bal:66]: Noticed that the same doc comment
`Creates a record instance in NetSuite according to the given detail`

is used for most of the remote functions.

Shall we add proper doc comments for remote operations as they will be visible to the choreo users?

File: netsuite/commonUtils.bal [hunk 4]
```diff
@@ -110,7 +114,7 @@ isolated function getUpdateXMLBodyWithParentElement(string subElements) returns
     </soapenv:Envelope>`;
 }
 
-isolated function buildAddRecord(NewRecordType recordType, RecordCoreType recordCoreType, ConnectionConfig config) 
+isolated function getPayloadAddOp(NewRecordType recordType, RecordCoreType recordCoreType, ConnectionConfig config)
```
Comment (LakshanSS) [inline netsuite/commonUtils.bal:117]: What is the meaning of `getPayloadAddOp`?

Shall we use a proper name for this?

File: netsuite/responseUtils.bal [hunk 5]
```diff
@@ -22,7 +22,7 @@ import ballerina/xmldata;
 
 xmlns "urn:core_2020_2.platform.webservices.netsuite.com" as platformCore;
 
-isolated function getCreateResponse(http:Response response) returns @tainted RecordAddResponse|error {
+isolated function getResponseCreationOp(http:Response response) returns @tainted RecordAddResponse|error {
```
Comment (LakshanSS) [inline netsuite/responseUtils.bal:25]: Shall we use a proper meaningful name than `getResponseCreationOp`?

File: netsuite/commonTypes.bal [hunk 6]
```diff
@@ -44,3 +44,50 @@ public type Address record {
...
+public type MultiSelectCustomFieldRef record {|
+    string internalId;
+    string scriptId?;
+    ListOrRecordRef[] value;
+|};
+
+public type ListOrRecordRef record {
+    string recordName;
+    string internalId;
+};
```
Comment (LakshanSS) [inline netsuite/commonTypes.bal:92]: All the public records should contain doc comments.
They are needed to auto-generate the API Documentation in ballerina-central

File: netsuite/commonUtils.bal [hunk 7]
```diff
@@ -345,3 +366,142 @@ isolated function getXMLBodyForGetSavedSearchIDs(string searchType) returns stri
     return string `<soapenv:Body><urn:getSavedSearch><record searchType="${searchType}"/></urn:getSavedSearch>
         </soapenv:Body></soapenv:Envelope>`;
 }
+
+isolated function getCustomElementList(CustomFieldList customFieldList, string namespace) returns string|error {
+    string fieldsWithParentElement = string `<${namespace}:customFieldList xmlns:platformCore="urn:core_2020_2.platform.webservices.netsuite.com">`;
+    foreach var item in customFieldList.customFields {
+        if(item is LongCustomFieldRef) {
```
Comment (LakshanSS) [inline netsuite/commonUtils.bal:373]: Formatting issue
```suggestion
        if (item is LongCustomFieldRef) {
```

File: netsuite/commonUtils.bal [hunk 8]
```diff
@@ -345,3 +366,142 @@ isolated function getXMLBodyForGetSavedSearchIDs(string searchType) returns stri
...
+
+isolated function extractListOrRecordRefFromXML(xml element) returns ListOrRecordRef {
+    return {
+        recordName: extractStringFromXML(element/<name>/*),
+        internalId: extractStringFromXML(element.internalId)
+    };
+}
+
+
+
```
Comment (LakshanSS) [inline netsuite/commonUtils.bal:471]: Shall we remove unnecessary newlines?

File: netsuite/tests/test.bal [hunk 9]
```diff
@@ -848,3 +850,238 @@ function testAccountGetOperation() {
...
+            ]
+        }
+    };
+    var output = netsuiteClient->addNewVendorBill(vendorBill);
+    if (output is RecordAddResponse) {
+        log:printInfo(output.toString());
+        vendorBillId = output.internalId;
+    } else {
+        test:assertFail(output.toString());
+    }
```
Comment (LakshanSS) [inline netsuite/tests/test.bal:1014]: Here we can get rid of `var` and use `check`.
So we dont need the `if else`

```suggestion
    RecordAddResponse output = check netsuiteClient->addNewVendorBill(vendorBill);
```

File: netsuite/tests/test.bal [hunk 10]
```diff
@@ -848,3 +850,238 @@ function testAccountGetOperation() {
...
+    RecordInfo recordInfo = {
+        recordInternalId : vendorBillId,
+        recordType: "vendorBill"
+    };
+    VendorBill|error output = netsuiteClient->getVendorBillRecord(recordInfo);
+    if (output is Vendor) {
+        log:printInfo(output.toString());
+    } else {
+        test:assertFail(output.toString());
+    }
```
Comment (LakshanSS) [inline netsuite/tests/test.bal:1029]:
```suggestion
    VendorBill output = check netsuiteClient->getVendorBillRecord(recordInfo);
```

File: netsuite/vendor.bal (hunk 2 above)
Comment (SanduDS) [reply netsuite/vendor.bal:162]: This is used in NetSuite xml result to ballerina record manual mapping as still XML to ballerina record mapping is not fully supported. NetSuite does not provide values for all the fields in the record for every request. There can be missing XML elements. it depends on the business logic and the values of the NetSuite record. Therefore some records might have values for some fields, but they might not be in another record. If we use 'check' we will get an error every time when Netsuite doesn't provide values for all fields. As we use manual mapping with xml payloads, we can't get rid of if blocks. But when we got full XML to record mapping support, these all can be removed.

Comment (LakshanSS) [reply netsuite/vendor.bal:162]: +1

File: netsuite/client.bal [hunk 11]
```diff
@@ -40,91 +40,115 @@ public isolated client class Client {
         self.basicClient = check new (config.baseURL + NETSUITE_ENDPOINT, httpClientConfig);
     }
 
-    # Creates a record instance in NetSuite according to the given detail
+    # Creates a new vendor record instance in NetSuite according to the given detail
```
Comment (abeykoon) [inline netsuite/client.bal:43]: Needs a full stop mark at the end of the description.

File: netsuite/commonUtils.bal [hunk 12]
```diff
@@ -110,7 +114,7 @@ isolated function getUpdateXMLBodyWithParentElement(string subElements) returns
     </soapenv:Envelope>`;
 }
 
-isolated function buildAddRecord(NewRecordType recordType, RecordCoreType recordCoreType, ConnectionConfig config) 
+isolated function getPayloadForAddOperation(NewRecordType recordType, RecordCoreType recordCoreType, ConnectionConfig config)
```
Comment (abeykoon) [inline netsuite/commonUtils.bal:117]: `buildAddOperationPayload`

**Review by abeykoon — APPROVED** [review]

**Pull request merged**